src/main/groovy/edu/oregonstate/mist/cardsapi/resources/CardsResource.groovy

```diff
@@ -65,4 +83,153 @@ class CardsResource extends Resource {
+            if(invalidTypes) {
+                return badRequest("Invalid types: \'${invalidTypes.join("\', \'")}\'. " +
+                        "Valid types are: " +
+                        "skill, attack, power, status, curse").build()
+            }
+        }
+
+        if(!colors) {
+            colors = validColors
+        } else {
+            List<String> invalidColors = colors - validColors
+            if(invalidColors) {
+                return badRequest("Invalid colors: \'${invalidColors.join("\', \'")}\'. " +
+                        "Valid colors are: " +
+                        "red, green, blue, colorless.").build()
+            }
+        }
+
+        if(!rarities) {
+            rarities = validRarities
+        } else {
+            List<String> invalidRarities = rarities - validRarities
+            if(invalidRarities) {
+                return badRequest("Invalid rarities: \'${invalidRarities.join("\', \'")}\'. " +
+                        "Valid rarities are: " +
+                        "basic, common, uncommon, rare").build()
+            }
+        }
+
+        if(!(name.or("")).matches(regEx)) {
+            return badRequest("Invalid name: \'" + name.get() +
+                    "\'. Name must match pattern: " +
+                    regEx).build()
+        }
+
+        for(int i = 0; i < keywords.size(); i++) {
+            if(!keywords[i].matches(regEx)) {
+                return badRequest("Invalid keyword: \'" + keywords[i] +
+                        "\'. All keywords " +
+                        "must match pattern: " + regEx).build()
+            }
+        }
+
+        List<Card> cards = cardFluent.getCards(types, name.orNull(), colors, rarities,
+                energyMin.or(0), energyMax.or(999),
+                keywords, number.or(10), isRandom.or(true))
+
+        ResultObject cardResult = cardsResult(cards)
+        ok(cardResult).build()
+    }
+
+    @POST
+    @Path('')
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    Response postCard (@Valid Card newCard) {
+
+        Response response = cardValidator(newCard)
+        if(response) {
+            return response
+        }
+
+        Integer id = cardDAO.getNextId()
+        cardDAO.postCard(
+                id,
+                newCard.type,
+                newCard.name,
+                newCard.color,
+                newCard.rarity,
+                newCard.energy,
+                newCard.description)
+
+        Card card = cardDAO.getCardById(id)
+        ResultObject cardResult = cardsResult(card)
+        created(cardResult).build()
+    }
+
+    @PUT
+    @Path('{id}')
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    Response putCard(@PathParam('id') IntParam id, @Valid Card updateCard) {
+
+        if(cardDAO.cardExists(id.get()) != 1) {
+            return notFound().build()
+        }
+        Response response = cardValidator(updateCard)
```

Comment:
could you name this variable something like `badResponse` or something? It will be easier to know that if that response contains something, a bad request is returned to the client.